// umi-build-dev/lib/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.EXT_LIST = exports.PLACEHOLDER_HISTORY_MODIFIER = exports.PLACEHOLDER_ROUTER = exports.PLACEHOLDER_RENDER = exports.PLACEHOLDER_ROUTES_MODIFIER = exports.PLACEHOLDER_ROUTER_MODIFIER = exports.PLACEHOLDER_IMPORT = exports.CONFIG_FILES = void 0;
const CONFIG_FILES = ['.umirc.js', 'config/config.js'];
exports.CONFIG_FILES = CONFIG_FILES;
const PLACEHOLDER_IMPORT = '<%= IMPORT %>';
exports.PLACEHOLDER_IMPORT = PLACEHOLDER_IMPORT;
const PLACEHOLDER_ROUTER_MODIFIER = '<%= ROUTER_MODIFIER %>';
exports.PLACEHOLDER_ROUTER_MODIFIER = PLACEHOLDER_ROUTER_MODIFIER;
const PLACEHOLDER_ROUTES_MODIFIER = '<%= ROUTES_MODIFIER %>';
exports.PLACEHOLDER_ROUTES_MODIFIER = PLACEHOLDER_ROUTES_MODIFIER;
const PLACEHOLDER_RENDER = '<%= RENDER %>';
exports.PLACEHOLDER_RENDER = PLACEHOLDER_RENDER;
const PLACEHOLDER_ROUTER = '<%= ROUTER %>';
exports.PLACEHOLDER_ROUTER = PLACEHOLDER_ROUTER;
const PLACEHOLDER_HISTORY_MODIFIER = '<%= HISTORY_MODIFIER %>';
exports.PLACEHOLDER_HISTORY_MODIFIER = PLACEHOLDER_HISTORY_MODIFIER;
const EXT_LIST = ['.js', '.jsx', '.ts', '.tsx'];
exports.EXT_LIST = EXT_LIST;
