// umi-plugin-dva/lib/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getModel = getModel;
exports.default = _default;

var _fs = require("fs");

var _path = require("path");

var _globby = _interopRequireDefault(require("globby"));

var _lodash = _interopRequireDefault(require("lodash.uniq"));

var _pathIsRoot = _interopRequireDefault(require("path-is-root"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; var ownKeys = Object.keys(source); if (typeof Object.getOwnPropertySymbols === 'function') { ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function (sym) { return Object.getOwnPropertyDescriptor(source, sym).enumerable; })); } ownKeys.forEach(function (key) { _defineProperty(target, key, source[key]); }); } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function getModel(cwd, api) {
  const config = api.service.config;
  const winPath = api.utils.winPath;
  const modelJSPath = (0, _path.join)(cwd, 'model.js');

  if ((0, _fs.existsSync)(modelJSPath)) {
    return [winPath(modelJSPath)];
  }

  const modelJSXPath = (0, _path.join)(cwd, 'model.jsx');

  if ((0, _fs.existsSync)(modelJSXPath)) {
    return [winPath(modelJSXPath)];
  }

  const modelTSPath = (0, _path.join)(cwd, 'model.ts');

  if ((0, _fs.existsSync)(modelTSPath)) {
    return [winPath(modelTSPath)];
  }

  const modelTSXPath = (0, _path.join)(cwd, 'model.tsx');

  if ((0, _fs.existsSync)(modelTSXPath)) {
    return [winPath(modelTSXPath)];
  }

  return _globby.default.sync(`./${config.singular ? 'model' : 'models'}/**/*.{ts,tsx,js,jsx}`, {
    cwd
  }).filter(p => !p.endsWith('.d.ts') && !p.endsWith('.test.js') && !p.endsWith('.test.jsx') && !p.endsWith('.test.ts') && !p.endsWith('.test.tsx')).map(p => winPath((0, _path.join)(cwd, p)));
}

function _default(api, opts = {}) {
  const _api$placeholder = api.placeholder,
        RENDER = _api$placeholder.RENDER,
        ROUTER_MODIFIER = _api$placeholder.ROUTER_MODIFIER,
        IMPORT = _api$placeholder.IMPORT;
  const _api$service = api.service,
        paths = _api$service.paths,
        config = _api$service.config;
  const winPath = api.utils.winPath;
  const dvaContainerPath = (0, _path.join)(paths.absTmpDirPath, 'DvaContainer.js');
  const isProduction = process.env.NODE_ENV === 'production';
  const shouldImportDynamic = isProduction && !config.disableDynamicImport;

  function getDvaJS() {
    if ((0, _fs.existsSync)((0, _path.join)(paths.absSrcPath, 'dva.js'))) {
      return winPath((0, _path.join)(paths.absSrcPath, 'dva.js'));
    }

    if ((0, _fs.existsSync)((0, _path.join)(paths.absSrcPath, 'dva.ts'))) {
      return winPath((0, _path.join)(paths.absSrcPath, 'dva.ts'));
    }
  }

  function endWithSlash(path) {
    return path.slice(-1) !== '/' ? `${path}/` : path;
  }

  function isPagesPath(path) {
    return endWithSlash(winPath(path)) === endWithSlash(winPath(paths.absPagesPath));
  }

  function isSrcPath(path) {
    return endWithSlash(winPath(path)) === endWithSlash(winPath(paths.absSrcPath));
  }

  function getModelsWithRoutes(routes) {
    return routes.reduce((memo, route) => {
      if (route.redirect) {
        return memo;
      }

      return [...memo, ...getPageModels((0, _path.join)(paths.cwd, route.component)), ...(route.routes ? getModelsWithRoutes(route.routes) : [])];
    }, []);
  }

  function getGlobalModels() {
    let models = getModel(paths.absSrcPath, api);

    if (!shouldImportDynamic) {
      // dev 模式下还需要额外载入 page 路由的 models 文件
      models = [...models, ...getModelsWithRoutes(api.service.routes)]; // 去重

      models = (0, _lodash.default)(models);
    }

    return models;
  }

  function getPageModels(cwd) {
    let models = [];

    while (!isPagesPath(cwd) && !isSrcPath(cwd) && !(0, _pathIsRoot.default)(cwd)) {
      models = models.concat(getModel(cwd, api));
      cwd = (0, _path.dirname)(cwd);
    }

    return models;
  }

  function getModelName(model) {
    const modelArr = winPath(model).split('/');
    return modelArr[modelArr.length - 1];
  }

  function exclude(models, excludes) {
    return models.filter(model => {
      var _iteratorNormalCompletion = true;
      var _didIteratorError = false;
      var _iteratorError = undefined;

      try {
        for (var _iterator = excludes[Symbol.iterator](), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
          const exclude = _step.value;

          if (typeof exclude === 'function' && exclude(getModelName(model))) {
            return false;
          }

          if (exclude instanceof RegExp && exclude.test(getModelName(model))) {
            return false;
          }
        }
      } catch (err) {
        _didIteratorError = true;
        _iteratorError = err;
      } finally {
        try {
          if (!_iteratorNormalCompletion && _iterator.return != null) {
            _iterator.return();
          }
        } finally {
          if (_didIteratorError) {
            throw _iteratorError;
          }
        }
      }

      return true;
    });
  }

  function optsToArray(item) {
    if (!item) return [];

    if (Array.isArray(item)) {
      return item;
    } else {
      return [item];
    }
  }

  function getGlobalModelContent() {
    return exclude(getGlobalModels(), optsToArray(opts.exclude)).map(path => `
    app.model({ ...(require('${path}').default) });
  `.trim()).join('\r\n');
  }

  function getPluginContent() {
    const pluginPaths = _globby.default.sync('plugins/**/*.{js,ts}', {
      cwd: paths.absSrcPath
    });

    const ret = pluginPaths.map(path => `
app.use(require('../../${path}').default);
  `.trim());

    if (opts.immer) {
      ret.push(`
app.use(require('${winPath(require.resolve('dva-immer'))}').default());
      `.trim());
    }

    return ret.join('\r\n');
  }

  function stripFirstSlash(path) {
    if (path.charAt(0) === '/') {
      return path.slice(1);
    } else {
      return path;
    }
  }

  function chunkName(path) {
    return stripFirstSlash(winPath(path).replace(winPath(paths.cwd), '')).replace(/\//g, '__');
  }

  api.register('generateFiles', () => {
    const tpl = (0, _path.join)(__dirname, '../template/DvaContainer.js');
    let tplContent = (0, _fs.readFileSync)(tpl, 'utf-8');
    const dvaJS = getDvaJS();

    if (dvaJS) {
      tplContent = tplContent.replace('<%= ExtendDvaConfig %>', `
...((require('${dvaJS}').config || (() => ({})))()),
        `.trim()); //         .replace('<%= EnhanceApp %>', `
      // app = (require('${dvaJS}').enhance || (app => app))(app);
      //         `.trim());
    }

    tplContent = tplContent.replace('<%= ExtendDvaConfig %>', '').replace('<%= EnhanceApp %>', '').replace('<%= RegisterPlugins %>', getPluginContent()).replace('<%= RegisterModels %>', getGlobalModelContent());
    (0, _fs.writeFileSync)(dvaContainerPath, tplContent, 'utf-8');
  });
  api.register('modifyRouterFile', ({
    memo
  }) => {
    return memo.replace(IMPORT, `
import { routerRedux } from 'dva/router';
${shouldImportDynamic ? `import _dvaDynamic from 'dva/dynamic';` : ''}
${IMPORT}
      `.trim()).replace(ROUTER_MODIFIER, `
const { ConnectedRouter } = routerRedux;
Router = ConnectedRouter;
${ROUTER_MODIFIER}
      `.trim());
  });

  if (shouldImportDynamic) {
    api.register('modifyRouteComponent', ({
      memo,
      args
    }) => {
      const pageJSFile = args.pageJSFile,
            webpackChunkName = args.webpackChunkName;

      if (!webpackChunkName) {
        return memo;
      }

      let ret = `
_dvaDynamic({
  <%= MODELS %>
  component: () => import(/* webpackChunkName: '${webpackChunkName}' */'${pageJSFile}'),
})
      `.trim();
      const models = getPageModels((0, _path.join)(paths.absTmpDirPath, pageJSFile));

      if (models && models.length) {
        ret = ret.replace('<%= MODELS %>', `
app: window.g_app,
models: () => [
  ${models.map(model => `import(/* webpackChunkName: '${chunkName(model)}' */'${model}')`).join(',\r\n  ')}
],
      `.trim());
      }

      return ret.replace('<%= MODELS %>', '');
    });
  }

  api.register('modifyEntryFile', ({
    memo
  }) => {
    const dvaRender = api.service.applyPlugins('modifyDvaRender', {
      initialValue: `
ReactDOM.render(React.createElement(
  DvaContainer,
  null,
  React.createElement(require('./router').default)
), document.getElementById('root'));
`.trim()
    });
    return memo.replace(RENDER, `
const DvaContainer = require('./DvaContainer').default;
${dvaRender}
`.trim());
  });
  api.register('modifyAFWebpackOpts', ({
    memo
  }) => {
    memo.alias = _objectSpread({}, memo.alias, {
      dva: (0, _path.dirname)(require.resolve('dva/package')),
      'dva-loading': require.resolve('dva-loading'),
      'path-to-regexp': require.resolve('path-to-regexp'),
      'object-assign': require.resolve('object-assign')
    }, opts.immer ? {
      immer: require.resolve('immer')
    } : {});
    return memo;
  });
  api.register('modifyPageWatchers', ({
    memo
  }) => {
    return [...memo, (0, _path.join)(paths.absSrcPath, 'models'), (0, _path.join)(paths.absSrcPath, 'plugins'), (0, _path.join)(paths.absSrcPath, 'model.js'), (0, _path.join)(paths.absSrcPath, 'model.jsx'), (0, _path.join)(paths.absSrcPath, 'model.ts'), (0, _path.join)(paths.absSrcPath, 'model.tsx'), (0, _path.join)(paths.absSrcPath, 'dva.js'), (0, _path.join)(paths.absSrcPath, 'dva.ts')];
  });
}
